### PR TITLE
A couple of things

### PR DESCRIPTION
**Pull request by tamland**
1. Added defaults for the `Plugin`  constructor. I don't see why anybody would need to override these, but I left them for backwards comparability..
2. Added a `route_for` method to determine where an url routes. A use-case for this can be seen here: https://github.com/takoi/xbmc-addon-nrk/blob/master/default.py#L63 Basically, it replaces the entire playable-or-directory logic by using route_for to find out whether it routes to the play node or not.

### CODE REVIEW

File: xbmcswift2/plugin.py [hunk 1]
```diff
@@ -49,13 +50,16 @@ class Plugin(XBMCMixin):
                      builtin ``__file__`` variable can used.
     '''
 
-    def __init__(self, name, addon_id, filepath):
-        self._name = name
+    def __init__(self, name='', addon_id='', filepath=__main__.__file__):
+        if addon_id:
+            self._addon = xbmcaddon.Addon(id=addon_id)
+        else:
+            self._addon = xbmcaddon.Addon()
```
Comment (jbeluch) [inline xbmcswift2/plugin.py:57]: I believe the "id" argument is required?

File: xbmcswift2/plugin.py [hunk 2]
```diff
@@ -49,13 +50,16 @@ class Plugin(XBMCMixin):
                      builtin ``__file__`` variable can used.
     '''
 
-    def __init__(self, name, addon_id, filepath):
-        self._name = name
+    def __init__(self, name='', addon_id='', filepath=__main__.__file__):
```
Comment (jbeluch) [inline xbmcswift2/plugin.py:53]: While I like the idea of provided a default filepath, we cannot rely on `__main__.__file__`. When you run an addon on the command line using `xbmcswift2 run`, the `__main__.__file__` will be the xbmcswift2 binary, not the addon's main python file.

File: xbmcswift2/plugin.py (hunk 1 above)
Comment (tamland) [inline xbmcswift2/plugin.py:57]: No, xbmc will inject the proper id

File: xbmcswift2/plugin.py (hunk 2 above)
Comment (tamland) [inline xbmcswift2/plugin.py:53]: Didn't realize that. I could never get the cli to work, so I didn't test it.

But if I understand this correctly now: `filepath` is only used to get to the resources folder, correct?
The currently, if you put your file somewhere else, say in resources/lib like many people do, it won't work. The doc doesn't say anything about this but only that it should be the path to the addon.py. Since everything else requires cwd to be root of the addon folder, why not just use cwd for the resource folder too?

Comment (jbeluch) [inline xbmcswift2/plugin.py:53]: You are correct, filepath is only used to find the resources folder. And `xbmcswift2 run` requires the cwd to be the root of the addon so this could definitely be changed. I created #6 to keep track.

**Comment by jbeluch** [comment]
I'm going to work on merging some of this now. Thanks for the contribution!

Aso, it would be easier (for me, not you ;) in the future if you created separate topic branches for different fixes/suggestions. Then each PR would only have commits/comments related to that change and not have 3 different fixes all mixed together.

**Comment by tamland** [comment]
Sure, but I can at least rebase down to the two if you want

**Comment by jbeluch** [comment]
Btw, I noticed that in your link from the first post (https://github.com/takoi/xbmc-addon-nrk/blob/master/default.py#L63), you are using xbmcswift not xbmcswift2. xbmcswift is the older version.

**Comment by tamland** [comment]
I know, I haven't upgraded yet. Why?

**Comment by jbeluch** [comment]
You said you were having CLI issues with xbmcswift so I was curious if you were using the new version or not, that's all.

**Comment by tamland** [comment]
Oh. No, I think the cli problems has to do with how I use the xbmc api directly. You see I wrote that plugin before adding xbmcswift. Anyway, I see you've added the default parameters, so I guess 77fec85 is the only thing left here.

**Comment by jbeluch** [comment]
Sorry this took so long to reply :)

I'm not sure the route_for method belongs in xbmcswift2. I don't think its a good idea to match the xbmcswift2 url patterns to the website url patterns. I think they should be decoupled. For normal usage, I don't think route_for is necessary.

A lot of people simply subclass Plugin to add custom methods they like to use. This would probably work well in your case if you continue to use route_for. Thanks for the other contributions!

**Comment by tamland** [comment]
"match to the website url patterns" were did you get that from? That's not what it does, that's just something I do in that specific addon because the site use so many different urls and patters to point to the same page, doing something else would be a hell maintaining.

Maybe you haven't understood what it does so let me try to explain again: `url_for` maps endpoint and arguments to a url right. Well, `route_for` simply maps url to the endpoint. It's the reverse of `for_url`. This does not "couple" anything. It doesn't even accept urls, only internal, relative and valid paths. It's a common function to have in web frameworks, the same frameworks you have modeled xbmcswift after.

**Comment by jbeluch** [comment]
I understand what `route_for` does. I don't think its necessary because at some point for a given list item, you have to build the url by calling `url_for`. This requires you to pass an explicit endpoint, therefore making `route_for` unnecessary.

I looked at your example use case but perhaps I misunderstood. Since you seemed to be matching website urls to your patterns, you didn't necessarily know which endpoint to use hence the need for `route_for`. In a normal scenario, where you are building list items from dicts and calling `url_for`, I don't see `route_for` being necessary.

If I'm being dense and not fully understanding when it would be used or why it's necessary, please let me know.

**Comment by tamland** [comment]
You said it. It useful when you don't know the endpoint or don't want to fix it. You do what you want, but I thought xbmcswift were supposed to be a collection of commonly used functionality when developing plugins, not a one size fits all design pattern.

Either way, you should really consider to some times use a more dynamic approach in your own plugins. Not passing explicit endpoints is a great way to present a dynamic tree while not getting tied to a specific page.
